index.js

```diff
@@ -135,48 +138,65 @@ Filter.prototype.build = function() {
-        var outputFilePath = outputPath;
-
-        plugin._logger.debug('[operation:%s] %s', operation, relativePath);
-
-        switch (operation) {
-          case 'mkdir': {
-            instrumentation.mkdir++;
-            return fs.mkdirSync(outputPath);
-          } case 'rmdir': {
-            instrumentation.rmdir++;
-            return fs.rmdirSync(outputPath);
-          } case 'unlink': {
-            instrumentation.unlink++;
-            return fs.unlinkSync(outputPath);
-          } case 'change': {
-            instrumentation.change++;
-            return plugin._handleFile(relativePath, srcDir, destDir, entry, outputFilePath, true, instrumentation);
-          } case 'create': {
-            instrumentation.create++;
-            return plugin._handleFile(relativePath, srcDir, destDir, entry, outputFilePath, false, instrumentation);
-          } default: {
-            instrumentation.other++;
+      return new Promise(function(resolve) {
+        var result = mapSeries(patches, function(patch) {
+          var operation = patch[0];
+          var relativePath = patch[1];
+          var entry = patch[2];
+          var outputPath = destDir + '/' + (plugin.getDestFilePath(relativePath, entry) || relativePath);
+          var outputFilePath = outputPath;
+
+          plugin._logger.debug('[operation:%s] %s', operation, relativePath);
+
+          switch (operation) {
+            case 'mkdir': {
+              instrumentation.mkdir++;
+              return fs.mkdirSync(outputPath);
+            } case 'rmdir': {
+              instrumentation.rmdir++;
+              return fs.rmdirSync(outputPath);
+            } case 'unlink': {
+              instrumentation.unlink++;
+              return fs.unlinkSync(outputPath);
+            } case 'change': {
+              instrumentation.change++;
+              var changeOperation = plugin._handleFile(relativePath, srcDir, destDir, entry, outputFilePath, true, instrumentation);
+              // will be undefined if this.canProcessFile() == false
+              if (plugin.async && changeOperation !== undefined) {
+                pendingWork.push(changeOperation);
+                return;
+              }
+              return changeOperation;
+            } case 'create': {
+              instrumentation.create++;
+              var createOperation = plugin._handleFile(relativePath, srcDir, destDir, entry, outputFilePath, false, instrumentation);
+              // will be undefined if this.canProcessFile() == false
+              if (plugin.async && createOperation !== undefined) {
+                pendingWork.push(createOperation);
+                return;
+              }
+              return createOperation;
+            } default: {
+              instrumentation.other++;
+            }
           }
-        }
+        });
+        resolve(result);
+      }).then(() => {
+        const worker = queue.async.asyncify((promise) => promise);
```

Comment:
This wont do any throttling, as the promise merely represents the value of the work. The costly operation should be run in this callback.